relentless/potential/pair.py

```diff
@@ -156,6 +157,213 @@ def _derivative(self, param, r, epsilon, sigma, **params):
             d = d.item()
         return d
 
+class Spline(PairPotential):
+    """Spline potential using Akima interpolation.
+
+    Parameters
+    ----------
+    types : array_like
+        List of types (A type must be a `str`).
+    num_knots : int
+        The number of knots with which to interpolate.
+    mode : `str`
+        Describes the type of design parameter used in the optimization. Setting the mode to 'value' uses the knot amplitudes,
+        while 'diff' uses the differences between neighboring knot amplitudes. (Defaults to 'diff').
+
+    Raises
+    ------
+    ValueError
+        If the number of knots is not an integer >= 2.
+    ValueError
+        If the mode is not set to either 'value' or 'diff'.
+
+    """
+    valid_modes = ('value','diff')
+
+    def __init__(self, types, num_knots, mode='diff'):
+        if isinstance(num_knots,int) and num_knots >= 2:
+            self._num_knots = num_knots
+        else:
+            raise ValueError('Number of spline knots must be an integer >= 2.')
+
+        if mode in self.valid_modes:
+            self._mode = mode
+        else:
+            raise ValueError('Invalid parameter mode, choose from: ' + ','.join(Spline.valid_modes))
+
+        super().__init__(types=types,
+                         params=['r-{}'.format(i) for i in range(num_knots)] + ['knot-{}'.format(i) for i in range(num_knots)])
+
+    def from_array(self, pair, r, u):
+        """Sets up the potential from arrays of knot positions and energies.
+
+        Parameters
+        ----------
+        pair : tuple
+            The type pair (i,j) for which to set up the potential.
+        r : array_like
+            The array of knot positions.
+        u : array_like
+            The array of energies at each knot position.
+
+        Raises
+        ------
+        ValueError
+            If the array of r values does not have the same length as the number of knots.
+        ValueError
+            If the array of u values does not have the same length as the number of knots.
+
+        """
+        # check that r and u have the right shape
+        if len(r) != self.num_knots:
+            raise ValueError('r must have the same length as the number of knots')
+        if len(u) != self.num_knots:
+            raise ValueError('u must have the same length as the number of knots')
+
+        # convert to r,knot form given the mode
+        rs = np.asarray(r, dtype=np.float64)
+        ks = np.asarray(u, dtype=np.float64)
+        if self.mode == 'diff':
+            # difference is next knot minus my knot, with last knot fixed at its current value
+            ks[:-1] -= ks[1:]
+
+        # make all knots variables, but hold all r and the last knot const
+        for i in range(self.num_knots):
+            self.coeff[pair]['r-{}'.format(i)] = core.Variable(rs[i],const=True)
+            self.coeff[pair]['knot-{}'.format(i)] = core.Variable(ks[i],const=(i==self.num_knots-1))
```

Comment:
Let's expand this so that if the knot is already a variable, we just set the value:
```py
ri = 'r-{}'.format(i)
if self.coeff[pair][ri] is None:
    self.coeff[pair][ri] = core.Variable(rs[i],const=True)
else:
    self.coeff[pair][ri].value = rs[i]
```

As a general point, it would be also be good to add a method that can spit back the knot variables
```py
def _knot_params(self, i):
    assert isinstance(i,int), "Knots are keyed by integers"
    return 'r-{}'.format(i), 'knot-{}'.format(i)
```
For use in the constructor and other methods. That way we can always rename the variables later without having to change much code.